win32ss/gdi/ntgdi/freetype.c

```diff
@@ -2994,38 +2993,103 @@ static unsigned int get_bezier_glyph_outline(FT_Outline *outline, unsigned int b
 static INT
 IntRequestFontSize(PDC dc, FT_Face face, LONG Width, LONG Height)
 {
-    FT_Size_RequestRec  req;
+    FT_Error error;
+    ULONG TargetHeight, TargetWidth, RealHeight;
 
-    if (Width < 0)
-        Width = -Width;
+    ASSERT(face != NULL);
+    ASSERT(face->height > 0);
+    ASSERT(face->units_per_EM > 0);
+
+    if (Height == 0)
+    {
+        Height = dc->ppdev->devinfo.lfDefaultFont.lfHeight;
+    }
 
     if (Height < 0)
     {
-        Height = -Height;
+        /* case (A): lfHeight is negative: the character height */
+        Height = -Height << 6;
     }
-    if (Height == 0)
+    else if (Height > 0)
     {
-        Height = dc->ppdev->devinfo.lfDefaultFont.lfHeight;
+        /* case (B): lfHeight is positive: the cell height */
+        Height <<= 6;
+        Height = (Height * face->units_per_EM) / face->height;
     }
-    if (Height == 0)
+
+    TargetHeight = (Height * face->height) / face->units_per_EM;
+    TargetWidth = Width = Width << 6;
+
+    if (Width && FT_IS_SCALABLE(face))
     {
-        Height = Width;
+        if (face_has_encoding(face, FT_ENCODING_SJIS) ||
+            face_has_encoding(face, FT_ENCODING_PRC) ||
+            face_has_encoding(face, FT_ENCODING_BIG5) ||
+            face_has_encoding(face, FT_ENCODING_WANSUNG) ||
+            face_has_encoding(face, FT_ENCODING_JOHAB))
+        {
+            /* Case (i). Likely, the face has many fullwidth characters.
+               The width value was same as fullwidth character width. */
+        }
+        else
+        {
+            /* Case (ii). It's a Latin-like simple face.
+               The width value was same as halfwidth character width. */
+            Width *= 2;
+            TargetWidth *= 2;
+        }
     }
 
-    if (Height < 1)
-        Height = 1;
+    error = FT_Set_Char_Size(face, Width, Height, 96, 96);
+    RealHeight = face->size->metrics.ascender - face->size->metrics.descender;
+
+    /* FIXME: Optimize speed */
+#define DELTA1 64
+#define DELTA2 1
+    if (TargetHeight < RealHeight)
+    {
+        do
+        {
+            Height -= DELTA1;
+            if (Width != 0)
+                Width -= MulDiv(DELTA1, TargetWidth, TargetHeight);
+            error = FT_Set_Char_Size(face, Width, Height, 96, 96);
```

Comment:
Should you break out of the loop in case this function returns failure?

Reply:
Yes, I agree. I'll fix it.

Reply:
I fixed it.